torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+            else:
+                raise RuntimeError("Output {} of the user-provided function does not require gradients."
+                                   " The outputs must be computed in a differentiable manner from the input"
+                                   " when running in strict mode.".format(i))
+
+def _autograd_grad(outputs, inputs, grad_outputs=None, create_graph=False, retain_graph=None):
+    # Version of autograd.grad that accepts `None` in outputs and do not compute gradients for them.
+    # This has the extra constraint that inputs has to be a tuple
+    if torch.is_tensor(outputs):
+        outputs = (outputs,)
+        grad_outputs = (grad_outputs,)
+    if grad_outputs is None:
+        grad_outputs = (None,) * len(outputs)
+
+    new_outputs = tuple()
+    new_grad_outputs = tuple()
+    for out, grad_out in zip(outputs, grad_outputs):
+        if out is not None and out.requires_grad:
+            new_outputs += (out,)
+            new_grad_outputs += (grad_out,)
+
+    if len(new_outputs) == 0:
+        # No differentiable output, we don't need to call the autograd engine
+        return (None,) * len(inputs)
+    else:
+        return torch.autograd.grad(new_outputs, inputs, new_grad_outputs, allow_unused=True,
+                                   create_graph=create_graph, retain_graph=retain_graph)
+
+# Public API
+
+def vjp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between a vector ``v`` and the Jacobian of
+    the given function at the point given by the inputs.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the output of ``func``. This argument is optional when
+            ``func``'s output contains a single element and it will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict if ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect we detect that some quantities
+            are independent of the element we want to differentiate it with. If ``False``, such quantity
+            will only be set to ``0`` which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
+        result (tuple of Tensor or Tensor): result of the dot product with the same shape
```

Comment:
It feels strange that we're using "shape" or "size" to refer to the "shape of a container and its tensor constituents", but I don't see a better way to describe it

Reply:
Yes, we use shape to refer to the nested structure of optional tuple + Tensor. Maybe I should use a word that does not have connotation as being a Tensor size/shape? like structure?

Reply:
I don't know what the right answer for this is. I'm fine with keeping it as "shape" for now, and then perhaps elaborating on it in the future or picking a new way to describe this.